immscore/md5digest.h

```diff
@@ -40,35 +40,49 @@ class Md5Digest
         static unsigned char bin_buffer[128 / 8];
         static char hex_buf[34] = {'\0'};
         static char tag_buf[4] = {'\0'};
+        size_t nmemb_read;
+        long offset = OFFSET;
+        char *cur = hex_buf;
+        int res, err;
 
         FILE *fp = fopen(filename.c_str(), "r");
         if (!fp)
-            return "bad_checksum";
+            goto failed;
 
-        long offset = OFFSET;
-        fseek(fp, TAGOFFSET, SEEK_END);
-        size_t nmemb_read = fread(tag_buf, 4, 1, fp);
+        if (fseek(fp, TAGOFFSET, SEEK_END) == -1)
+            goto failed;
+
+        nmemb_read = fread(tag_buf, 4, 1, fp);
         if (nmemb_read != 1)
-            return "bad_checksum";
+            goto failed;
 
         if (!strncmp(tag_buf, "TAG", 3))
             offset += TAGOFFSET;
 
-        int res = fseek(fp, offset, SEEK_END);
-
+        res = fseek(fp, offset, SEEK_END);
         if (res)
             rewind(fp);
 
-        int err = md5_stream(fp, NUMBLOCKS, bin_buffer);
-        fclose(fp);
+        err = md5_stream(fp, NUMBLOCKS, bin_buffer);
         if (err)
-            return "bad_checksum";
+            goto failed;
+
+        err = fclose(fp);
+        if (err)
+            goto failed;
 
-        char *cur = hex_buf;
         for (int i = 0; i < (128 / 4 / 2); ++i)
             cur += sprintf(cur, "%02x", bin_buffer[i]);
 
         return hex_buf;
+
+failed:
+        LOG(ERROR) << "Error hashing file \"" << filename << "\": " << error_string(errno) << endl;
+        if (fp)
```

Comment:
Instead of checking this, you can split this into two labels and use label fallthrough: "goto failed_fclose" vs. "goto failed". In both cases, you will log the error, but in only one case you will fclose. For example:

failed:
    fclose(fp);
failed_fclose:
    LOG_ERROR etc
failed_fclose:

Reply:
Wouldn't this lose the previous errno if fclose() failed?

Reply:
No; fp is a FILE *, so fclose can modify only what it points to, not the pointer value itself (it would need to be FILE ** for that to work). It can only return err, so fp would be unmodified.

Reply:
I don't understand. What I meant is that in your example above the original errno value (from the original error) would be overwritten by a failed fclose().

Reply:
Yes, sorry I misread your comment. I think it's a bit messy any way you slice it. Here's what I would probably recommend, but there are other ways too. Mostly, I think we should try to duplicate as little code as possible, always close when we need to, and log all errors (both fclose and not fclose) in the right order (non-fclose errors before fclose ones).

```
err = something();
if (err) {
    LOG err
    goto failed;
}
...
err = fclose(fp);
if (err)
    return
...
failed:
err = fclose(fp);
if (err)
    LOG error
```